File: Problem2.py
"""
I used binary search on the sorted citations list to find the H-Index. At each step, I compare the citation at the middle index with the number of papers that have at least that many citations. Based on the comparison, I adjust the search range. If there's no exact match, the H-Index is calculated as n - left.
Time Complexity: O(log n)
Space Complexity: O(1)
"""
class Solution:
    def hIndex(self, citations: List[int]) -> int:

        left, right = 0, len(citations) - 1
        n = len(citations)
        while left <= right:
            mid = left + (right - left) // 2
            if citations[mid] == n - mid:
                return n - mid
            elif citations[mid] < n - mid:
                left = mid + 1
            else:
                right = mid - 1
        return n - left

File: Problem2.js
// ## Problem2
// H-Index II (https://leetcode.com/problems/h-index-ii)

var hIndex = function(citations) {
    //Tc: log n
    //sc : constant
    const n = citations.length;
    let left = 0;
    let right = n - 1;
    
    while (left <= right) {
        const mid = Math.floor((left + right) / 2);
        const h = n - mid;
        
        if (citations[mid] >= h) {
            right = mid - 1;
        } else {
            left = mid + 1;
        }
    }
    
    return n - left;
};

console.log(hIndex([1,4,5,6]))

// var hIndex = function(citations) {
//     //TC : O(N)
//     //SC : O(1)
//     const n = citations.length;
//     let hIndex = 0;
    
//     for (let i = n - 1; i >= 0; i--) {
//         const currentH = n - i; // Number of papers from i to end
//         if (citations[i] >= currentH) {
//             hIndex = currentH;
//         } else {
//             break; // No need to check further
//         }
//     }
    
//     return hIndex;
// };